ast/doc.go

```diff
@@ -1,36 +1,6 @@
-// Copyright 2016 The OPA Authors.  All rights reserved.
+// Copyright 2024 The OPA Authors.  All rights reserved.
 // Use of this source code is governed by an Apache2
 // license that can be found in the LICENSE file.
 
-// Package ast declares Rego syntax tree types and also includes a parser and compiler for preparing policies for execution in the policy engine.
-//
-// Rego policies are defined using a relatively small set of types: modules, package and import declarations, rules, expressions, and terms. At their core, policies consist of rules that are defined by one or more expressions over documents available to the policy engine. The expressions are defined by intrinsic values (terms) such as strings, objects, variables, etc.
-//
-// Rego policies are typically defined in text files and then parsed and compiled by the policy engine at runtime. The parsing stage takes the text or string representation of the policy and converts it into an abstract syntax tree (AST) that consists of the types mentioned above. The AST is organized as follows:
-//
-//	Module
-//	 |
-//	 +--- Package (Reference)
-//	 |
-//	 +--- Imports
-//	 |     |
-//	 |     +--- Import (Term)
-//	 |
-//	 +--- Rules
-//	       |
-//	       +--- Rule
-//	             |
-//	             +--- Head
-//	             |     |
-//	             |     +--- Name (Variable)
-//	             |     |
-//	             |     +--- Key (Term)
-//	             |     |
-//	             |     +--- Value (Term)
-//	             |
-//	             +--- Body
-//	                   |
-//	                   +--- Expression (Term | Terms | Variable Declaration)
-//
-// At query time, the policy engine expects policies to have been compiled. The compilation stage takes one or more modules and compiles them into a format that the policy engine supports.
+// Deprecated: Use [github.com/open-policy-agent/opa/v1/ast] instead.
```

Comment:
This will put a deprecation message at the top of the doc page for the package:

<img width="1550" alt="Screenshot 2024-12-02 at 17 18 57" src="https://github.com/user-attachments/assets/b79f9e65-8d9c-413a-93f6-49f8b2b80bcd">

and it will cause linters (such as staticcheck used in OPA) to warn about usage of deprecated functions and types when those are referenced.

However, it won't put deprecated notices on anything inside the deprecated package in the generated docs. For this, we need to explicitly tag each type, const, var, and function. We could brute force it by just adding a generic

```
// 
// Deprecated: Use [github.com/open-policy-agent/opa/v1] instead.
```

everywhere. But before doing that to thousands of places, we should determine if it's necessary.

Reply:
Can web perhaps add some words, like "this package isn't going to go away but new projects are advised to start with v1"? If that's accurate. Just to avoid people panicking.

Reply:
You mean "web" as in what's generated from here, or as in the separate v0-migration guide we're working on for the docs?

How about:

```
Deprecated: This package is intended for older projects transitioning from OPA v0.x and will remain for the lifetime of OPA v1.x, but its use is not recommended. For newer features and behaviours, such as defaulting to the Rego v1 syntax, use [github.com/open-policy-agent/opa/v1/ast] instead.
```

Reply:
> Can web perhaps add some words

typing with gloves on in harsh weather, sorry 😅 "Can **we** perhaps add..." is what I had meant.

Reply:
I've updated all package deprecation notices to:

```
// Deprecated: This package is intended for older projects transitioning from OPA v0.x and will remain for the lifetime of OPA v1.x, but its use is not recommended.
// For newer features and behaviours, such as defaulting to the Rego v1 syntax, use the corresponding components in the [github.com/open-policy-agent/opa/v1] package instead.
// See https://www.openpolicyagent.org/docs/latest/v0-compatibility/ for more information.
```